src/OpenLoco/Map/TileManager.cpp

```diff
@@ -91,12 +92,45 @@ namespace OpenLoco::Map::TileManager
         TileManager::updateTilePointers();
     }
 
+    // Note: Must be past the last tile flag
+    static void markElementAsFree(TileElement& element)
+    {
+        element.setBaseZ(255);
+        if (element.next() == *_elementsEnd)
+        {
+            _elementsEnd--;
+        }
+    }
+
     // 0x00461760
     void removeElement(TileElement& element)
     {
-        registers regs;
-        regs.esi = X86Pointer(&element);
-        call(0x00461760, regs);
+        if (&element == *_F00158)
+        {
+            if (element.isLast())
+            {
+                *_F00158 = reinterpret_cast<TileElement*>(-1);
+            }
+        }
+
+        if (element.isLast())
+        {
+            auto* prev = element.prev();
+            prev->setLastFlag(true);
```

Comment:
This whole TileElement thing sounds like some linked list that's self-maintaining a bunch of things like which element is last - is there any way to move it to a `std` data structure eg vector?

Reply:
It's a bit of an odd structure that I've never found an equally good stl equivalent. It's designed to minimise allocations. It's not a linked list it's more like a vector with an associated map. The current design has the following properties: elements on a tile are contiguous; elements can be added/removed in O(1) time in best case; all elements are stored in contiguous memory;
The negatives are lots of adding/removing fragments memory. When memory runs low you need to defrag and reallocate which is an expensive operation. All pointers to memory are invalidated on addition/remove.
If you switched to multiple vectors then that's lots of allocations and you lose contiguous memory for all tile elements (slower load/save times). Linked list has the same issue.